ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java

```diff
@@ -461,6 +470,69 @@ private void cleanupOldState() {
     localCommands.ifPresent(lc -> lc.processLocalCommandFiles(serviceContext));
   }
 
+  public void cleanupOldStateDirectories(final KsqlConfig configWithApplicationServer) {
+    final String stateDir =
+        configWithApplicationServer
+            .getKsqlStreamConfigProps()
+            .getOrDefault(
+                StreamsConfig.STATE_DIR_CONFIG,
+                StreamsConfig.configDef().defaultValues().get(StreamsConfig.STATE_DIR_CONFIG))
+            .toString();
+
+    final Set<String> stateStoreNames =
+        ksqlEngine.getPersistentQueries()
+            .stream()
+            .map(PersistentQueryMetadata::getQueryApplicationId)
+            .collect(Collectors.toSet());
+    try {
+      Files.walkFileTree(
+          Paths.get(stateDir),
+          new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult visitFileFailed(final Path path, final IOException exc) {
+              log.error("Error cleaning up obsolete state directories \n", exc);
+              return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFile(final Path path, final BasicFileAttributes attrs) {
+              if (!stateStoreNames.contains(path.getFileName())) {
+                try {
+                  Files.delete(path);
+
+                  log.warn(
```

Comment:
Is `path.getFileName()` gonna be query names? Or would it be possible to be at the state store or even deeper level? Ditto below for `postVisitDirectory`.

Reply:
I don't know how this file visting system works like the first one did, but everything within the first level of the state store should match query names. Anything within those directories that _don't_ match current query names should be deleted.

So, if we hit `visitFile` for the first layer within the state store it should match the query application id, which is the list we've pulled above. Anything deeper I don't know -> makes me think we need to rethink this algo a bit. We want to delete recursively those directories in the first layer without checking to see if the files match